doc/aaa/ldap/hld_ldap.md

```diff
@@ -0,0 +1,514 @@
+
+#### LDAP BIND
+In order to use LDAP protocol is required to bind our switch to a LDAP server.
+That means, that after binding each client LDAP search to the Switch or SSH to the switch will be routed to the LDAP server.
+
+Flow of SSH login connection:
+Entry point  Command: ssh user_test@switch-133
+The user user_test will be searched in the LDAP server bound to the switch instead to the switch itself.
+The user authentication will be done by the LDAP server instead of the UNIX local authentication in the switch.
+
+How to do the binding configuration:
+The file that will contain all the info about binding is /etc/nslcd.conf.
+
+Those are the attributes required to set in order to bind client to server:
+
+binddn DN
+Specifies the distinguished name with which to bind to the directory server for lookups. The default is to bind anonymously.
+bindpw PASSWORD
+Specifies the credentials with which to bind. This option is only applicable when used with binddn above. If you set this option you should consider changing the permissions of the nslcd.conf file to only grant access to the root user.
+
+So, this file will contain all the LDAP configurations besides the login configuration that was described in the functional section above.
+Note: no restart is required when modifying PAM configuration. Only required to restart the NSLCD service after any modification in nslcd.conf file
+
+#### LDAP NSS
+LDAP can be used as an option in the Name Services Switch(NSS) configuration. The NSS configuration enables various programming APIs to use other sources than the default files (e.g., Use LDAP directory information instead of /etc/passwd for user and group information). User information includes uid, gid, and home directory.
+More details:
+NSS - A module-based system for controlling how various OS-level databases are assembled in memory. This includes (but is not limited to) passwd, group, shadow (this is important to note), and hosts. UID lookups use the passwd database, and GID lookups use the group database.
+
+#### Functional Flow
+LDAP Main flow – As mentioned in the LDAP authentication desc, the LDAP supports authentication by authenticating users via a remote server instead of locally (in the switch device).
+In high level the connection flow is the following:
+User will connect to a switch using ssh/login, the switch is an LDAP client (configured with LDAP feature - description of the configuration flow below), the client switch will “referred” the authentication of the user to the LDAP server binded, then the LDAP server will approve the authentication if the user & password match the LDAP server DB.
+And finally, the user will get approved and will be connected to the Switch.
```

Comment:
In our design, LDAP clients don't handle the deletion/addition/modification of users.
For now, we decided to remain aligned to the original design of the LDAP package, which means, that only the LDAP server is allowed to add/remove users and configure the permissions (mainly by setting to which group each user belongs.)

Note:
LDAP package has a tool named ldapmodify that can be used to modify the configuration in the server after passing LDAP bind authentication by remote, but we are not going to support this flow in the CLI.
 If a user likes to use this application it will be his responsibility.